openblas: +locking

ax3l opened the pull request:
BlasPP by ECP SLATE will fail to install by default (`spack install blaspp`) because
- the default BLAS installation in Spack is OpenBLAS
- BlasPP conflicts with `threads=none` for all recent OpenBLAS releases

Luckily, for OpenBLAS itself introduced a threadsafe compile option with 0.3.7+ aka `USE_LOCKING`:
```
   61 # If you want to build a single-threaded OpenBLAS, but expect to call this
   62 # from several concurrent threads in some other program, comment this in for
   63 # thread safety. (This is done automatically for USE_THREAD=1 , and should not
   64 # be necessary when USE_OPENMP=1)
   65 # USE_LOCKING = 1
```

According to my tests, with `spack install --test root blaspp`, this exactly addresses the issues in BlasPP tests.

It also seems to be a good option to set by default for OpenBLAS and users that do not need this safety net can always disable it.

cc @G-Ragghianti @mgates3

Follow-up to #20956

ax3l committed:
openblas: +locking

BlasPP by ECP SLATE will fail to install by default
(`spack install blaspp`) because:
- the default BLAS installation in Spack is OpenBLAS
- BlasPP conflicts with `threads=none` for all recent OpenBLAS releases

Luckily, for OpenBLAS itself introduced a threadsafe compile option
with 0.3.7+ aka `USE_LOCKING`:
```
   61 # If you want to build a single-threaded OpenBLAS, but expect to call this
   62 # from several concurrent threads in some other program, comment this in for
   63 # thread safety. (This is done automatically for USE_THREAD=1 , and should not
   64 # be necessary when USE_OPENMP=1)
   65 # USE_LOCKING = 1
```

According to my tests, with `spack install --test root blaspp`,
this exactly addresses the issues in BlasPP tests.

It also seems to be a good option to set by default for OpenBLAS and
users that do not need this safety net can always disable it.

G-Ragghianti commented:
Yes, I agree.  This seems to currently be the only way to fix this problem when installing blaspp. Is there a compelling reason to choose pthreads over openmp?

G-Ragghianti commented:
The alternative that I had attempted is something like this:

```
depends_on("blas");
depends_on("openblas threads=pthreads", when="^openblas");
```

But this doesn't work.

sethrj commented:
No, please don't! This is an issue for downstream parallel libraries like Trilinos that assume they manage the parallelism. Our particle transport codes use MPI and OpenMPI parallelism and make BLAS calls from each process/thread; with parallel BLAS that means extra threads stepping on each other.

~~If downstream apps rely on upstream parallel/thread options for a speed boost, then the spack recipes should specify those options.~~ I guess that could be argued either way -- and it's made difficult by BLAS being a virtual package, not  a concrete one. The reference netlib BLAS is unthreaded though...

ax3l commented:
> Is there a compelling reason to choose pthreads over openmp?

I picked pthread because openmp might be less portable as a default (looking at you, _AppleClang_ :eyes: )


> depends_on("openblas threads=pthreads", when="^openblas");

That's right, Spack unfortunately does not support this syntax.

> This is an issue for downstream parallel libraries like Trilinos

Ah dang it, thanks for the note. What do we do now? :scream: 
I need a somehow default-buildable BLAS++ for a specific geometry we use in WarpX #21787

Does someone know if `USE_THREAD` alone activates multi-threading in OpenBLAS? Or does it only add locks and mutexes?
I see conda-forge even activatates pthread & OpenMP [at the same time](https://github.com/conda-forge/openblas-feedstock/blob/c29a36d35a142d18edb8d28ca4aab5c216875cac/recipe/build.sh#L62).

I also found [notes like these](https://fossies.org/linux/OpenBLAS/Makefile.rule)
```
   61 # If you want to build a single-threaded OpenBLAS, but expect to call this
   62 # from several concurrent threads in some other program, comment this in for
   63 # thread safety. (This is done automatically for USE_THREAD=1 , and should not
   64 # be necessary when USE_OPENMP=1)
   65 # USE_LOCKING = 1
```
(We don't set `USE_LOCKING` yet.)

And for Power8:
```
   67 # If you're going to use this library with OpenMP, please comment it in.
   68 # This flag is always set for POWER8. Don't set USE_OPENMP = 0 if you're targeting POWER8.
   69 # USE_OPENMP = 1
```

There is more:
```
   79 # You can define the maximum number of threads. Basically it should be less
   80 # than or equal to the number of CPU threads. If you don't specify one, it's
   81 # automatically detected by the build system.
   82 # If SMT (aka. HT) is enabled on the system, it may or may not be beneficial to 
   83 # restrict NUM_THREADS to the number of physical cores. By default, the automatic 
   84 # detection includes logical CPUs, thus allowing the use of SMT.
   85 # Users may opt at runtime to use less than NUM_THREADS threads.
   86 #
   87 # Note for package maintainers: you can build OpenBLAS with a large NUM_THREADS
   88 # value (eg. 32-256) if you expect your users to use that many threads. Due to the way
   89 # some internal structures are allocated, using a large NUM_THREADS value has a RAM
   90 # footprint penalty, even if users reduce the actual number of threads at runtime.
   91 # NUM_THREADS = 24
   92 
   93 # If you have enabled USE_OPENMP and your application would call
   94 # OpenBLAS's calculation API from multiple threads, please comment this in.
   95 # This flag defines how many instances of OpenBLAS's calculation API can actually
   96 # run in parallel. If more than NUM_PARALLEL threads call OpenBLAS's calculation API,
   97 # they need to wait for the preceding API calls to finish or risk data corruption.
   98 # NUM_PARALLEL = 2
```

`NUM_THREADS`: `Users may opt at runtime to use less than NUM_THREADS threads.` <-- hey Trilinos? :)

alalazo commented:
For this one:
```
depends_on("blas");
depends_on("openblas threads=pthreads", when="^openblas");
```
did you try with the ASP based solver too (if that doesn't work I would consider it a bug)?

ax3l commented:
Random question: does the ASP solver also recognize conflicts, so he does not need to specify `threads=pthreads` or `threads=openmp`?

alalazo commented:
@ax3l Yep

ax3l commented:
w/o this PR, a general `spack solve blaspp` works.

The result is might be odd:
```
blaspp@2020.10.02%gcc@7.5.0+cuda~ipo+openmp+shared build_type=RelWithDebInfo cuda_arch=none arch=linux-ubuntu20.04-skylake
    ^cmake@3.19.4%gcc@7.5.0~doc+ncurses+openssl+ownlibs~qt arch=linux-ubuntu20.04-skylake
        ^ncurses@6.2%gcc@7.5.0~symlinks+termlib arch=linux-ubuntu20.04-skylake
            ^pkgconf@1.7.3%gcc@7.5.0 arch=linux-ubuntu20.04-skylake
        ^openssl@1.1.1j%gcc@7.5.0+systemcerts arch=linux-ubuntu20.04-skylake
            ^perl@5.32.1%gcc@7.5.0+cpanm+shared+threads arch=linux-ubuntu20.04-skylake
                ^berkeley-db@18.1.40%gcc@7.5.0 arch=linux-ubuntu20.04-skylake
                ^gdbm@1.18.1%gcc@7.5.0 arch=linux-ubuntu20.04-skylake
                    ^readline@8.0%gcc@7.5.0 arch=linux-ubuntu20.04-skylake
            ^zlib@1.2.11%gcc@7.5.0+optimize+pic+shared arch=linux-ubuntu20.04-skylake
    ^cuda@10.0.130%gcc@7.5.0 arch=linux-ubuntu20.04-skylake
    ^openblas@0.3.5%gcc@7.5.0~consistent_fpcsr~ilp64+pic+shared threads=none arch=linux-ubuntu20.04-skylake
```

It picks an old openblas instead of switching the thread variant.

Probably needs to be seen what is picked in context, e.g. with `py-warpx ^warpx +psatd dims=rz` #21787 (`==> Error: 'python'`)

G-Ragghianti commented:
> For this one:
> 
> ```
> depends_on("blas");
> depends_on("openblas threads=pthreads", when="^openblas");
> ```
> 
> did you try with the ASP based solver too (if that doesn't work I would consider it a bug)?

I haven't tried the ASP solver (I assume it is only available for testing through "spack solve"?).  What do I need to do to use it?  I see that clingo is required, but when I installed and loaded it within spack, I still get some error "Error: module 'clingo' has no attribute 'Control'".

sethrj commented:
Out of curiosity -- I hadn't heard of blaspp before this but had been hoping that a better API for BLAS existed outside of Trilinos' kind of sketchy Teuchos wrappers. (There are numerous issues with BLAS and LAPACK in mixed-compiler environments, including e.g. the Apple-distributed VecLib, and windows/intel toolchains.)  Is Blaspp intended to solve those problems?

For the reasons I mentioned above (working with apps that manage parallelism independently from the underlying linear algebra libraries) we wouldn't be able to use BLASpp if it requires BLAS-level multithreading. What is the underlying reason for the conflict with `openblas threads=none`? The package recipe just ays "tests will fail", but why would that be the case? Wouldn't the better solution to be a flag that informs `BLASpp` of thread support, add a `+multithreaded` (or whatever) variant to BLASpp, and update the BLASpp tests to support unthreaded BLAS installations?

ax3l commented:
> @G-Ragghianti: I haven't tried the ASP solver (I assume it is only available for testing through "spack solve"?). What do I need to do to use it? I see that clingo is required, but when I installed and loaded it within spack, I still get some error "Error: module 'clingo' has no attribute 'Control'".

@becker33 just told me the way again to try the new solver ([config ref](https://spack.readthedocs.io/en/latest/config_yaml.html#basic-settings)):
```bash
spack install clingo@master,spack
spack load --first clingo@master,spack

# default value  config:concretizer:original
spack config add config:concretizer:clingo
```

> BLAS++

@G-Ragghianti and @mgates3 are the perfect people to ask for your questions! It's part of ECP SLATE and early adopters that we are we use the nice C++ API in ECP WarpX :)

References for OpenBLAS issues: #20956 and (resolved/hidden) comments in #19311

G-Ragghianti commented:
> Out of curiosity -- I hadn't heard of blaspp before this but had been hoping that a better API for BLAS existed outside of Trilinos' kind of sketchy Teuchos wrappers. (There are numerous issues with BLAS and LAPACK in mixed-compiler environments, including e.g. the Apple-distributed VecLib, and windows/intel toolchains.) Is Blaspp intended to solve those problems?
> 
> For the reasons I mentioned above (working with apps that manage parallelism independently from the underlying linear algebra libraries) we wouldn't be able to use BLASpp if it requires BLAS-level multithreading. What is the underlying reason for the conflict with `openblas threads=none`? The package recipe just ays "tests will fail", but why would that be the case? Wouldn't the better solution to be a flag that informs `BLASpp` of thread support, add a `+multithreaded` (or whatever) variant to BLASpp, and update the BLASpp tests to support unthreaded BLAS installations?

Hi Seth.  @mgates3 is the lead on blaspp (and SLATE), so he can give you more details if needed.  I believe that a recent openblas update introduced something that isn't thread safe unless you configure it with pthreads or openmp support.  The reason blaspp can't automatically request thread support in openblas is that blaspp needs to be able to support other blas implementations though the "blas" virtual dependency.  This causes a problem with the current default spack concretizer (see my "depends_on" example above).  It may work with the new concretizer.  I'm about to test that.

sethrj commented:
> The package recipe just ays "tests will fail", but why would that be the case? Wouldn't the better solution to be a flag that informs `BLASpp` of thread support, add a `+multithreaded` (or whatever) variant to BLASpp, and update the BLASpp tests to support unthreaded BLAS installations?
> 
> Hi Seth. mgates3 is the lead on blaspp (and SLATE), so he can give you more details if needed. I believe that a recent openblas update introduced something that isn't thread safe unless you configure it with pthreads or openmp support. The reason blaspp can't automatically request thread support in openblas is that blaspp needs to be able to support other blas implementations though the "blas" virtual dependency. This causes a problem with the current default spack concretizer (see my "depends_on" example above). It may work with the new concretizer. I'm about to test that.

Thanks @G-Ragghianti . I understand the spack limitations on propagating the `threads` variant; but I'd like to know if it's possible to solve the underlying limitation preventing non-threaded OpenBLAS from working with BLASpp. It seems like that should be a standard use case. Even if it required a CMake configure toggle to disable thread support in BLASpp, the Spack problem would be solved -- you could easily set the necessary CMake flag based on the specs in `def configure`.

ax3l commented:
@mgates3 documented the issues [here](https://github.com/spack/spack/pull/19311#discussion_r554530892)

> @mgates3: We see an issue with OpenBLAS >= 0.3.6, verified through 0.3.13. It shows in our batched BLAS routines, which are doing regular BLAS inside an OpenMP parallel section. Regular BLAS routines are okay, as long as they aren't in an OpenMP parallel section. See https://github.com/xianyi/OpenBLAS/issues/3057

> @ax3l: Thank you for reporting this upstream, @mgates3.
Looks like we might be able to work-around with [locking](https://github.com/xianyi/OpenBLAS/issues/3057#issuecomment-757559901), let's see where you arrive at with upstream.

[More notes on virtual variants](https://github.com/spack/spack/pull/19311#discussion_r555298665), see #1712 and #6300.

@G-Ragghianti @mgates3 have you had the change to verify the OpenBLAS [USE_LOCKING](https://github.com/xianyi/OpenBLAS/issues/3057#issuecomment-757559901) option solves your observed test failures? We can easily add this to `openblas` in Spack and make it a requirement for OpenBLAS with BLAS++.

https://github.com/xianyi/OpenBLAS/blob/1caa44bea999e196918afc095d26ad418db4ba19/Makefile.rule#L61-L65

G-Ragghianti commented:
> @G-Ragghianti @mgates3 have you had the change to verify the OpenBLAS [USE_LOCKING](https://github.com/xianyi/OpenBLAS/issues/3057#issuecomment-757559901) option solves your observed test failures? We can easily add this to `openblas` in Spack and make it a requirement for OpenBLAS with BLAS++.

I haven't tested USE_LOCKING for openblas, but I will soon.

I can verify that the clingo concretizer does seem to correctly handle the following simultaneous dependencies:

```
depends_on('blas')
depends_on('openblas threads=openmp', when='^openblas')
```

This allows me to use a different blas provider with no complication.

ax3l commented:
Wuhu, thanks to the great test/check implementation in the `blaspp` package, I gave it a try already.
Indeed, we can enable `locking` for OpenBLAS 0.3.7+ to solve the test issues! :tada:

ax3l committed:
BLAS++: Adjust OpenBLAS Conflicts

Solve issues with newer OpenBLAS by requiring
`+locking` over none-default threading options.

ax3l commented:
I would even go more concrete to use
```
depends_on('openblas threads=openmp', when='+openmp ^openblas')
```
(Works well with clingo, does fail with the original dependency solver)

> I haven't tested USE_LOCKING for openblas, but I will soon.

Wuhu, thanks to the great test/check implementation in the `blaspp` package, I gave it a try already: `spack install --test root blaspp`. This is the issue I saw:
```
217 tests FAILED for batch-gemm
...
BLAS : Bad memory unallocation!
```
Indeed, enabling `USE_LOCKING=1` introduced with OpenBLAS 0.3.7+ solves the test issues! :tada:

I updated this PR to reflect the changes in OpenBLAS and BLAS++ and will wait for your and @mgates3's confirmation that this indeed is the solution.

I already did `spack install --test root blaspp` tests with OpenBLAS (Ubuntu 20.04/x86_64, GCC 10.2, original concretizer):

### 0.3.6

- [x] `spack install --test root blaspp ^openblas@0.3.6`
Spack abort by default with conflict (original concretizer).
`USE_LOCKING` came in [OpenBLAS 0.3.7 according to their docs](https://github.com/xianyi/OpenBLAS/blob/v0.3.7/Makefile.rule#L61-L65), so I added a new specific conflict

- [x] `spack install --test root blaspp ^openblas@0.3.6 threads=openmp`
BLAS++ 2020.10.02 (latest tag) CMake issue:
```
     30       BLAS library not found.
  >> 31    CMake Error at CMakeLists.txt:295 (message):
     32      BLAS++ requires a BLAS library and none was found.  Ensure that it is
     33      accessible in environment variables $CPATH, $LIBRARY_PATH, and
     34      $LD_LIBRARY_PATH.
```

### 0.3.7

- [x] `spack install --test root blaspp ^openblas@0.3.7`
same CMake error as with OpenBLAS 0.3.6

Otherwise this _should_ work because of the `+locking` default that came in this OpenBLAS release.

- [x] `spack install --test root blaspp ^openblas@0.3.7 threads=openmp`
same CMake error

- [ ] Do we want to simply add a conflict for selected old OpenBLAS releases in BLAS++'s `package.py`? Maybe that's also solved with newer BLAS++ releases that reworked the CMake, but the nwest release is still 2020.10.02 (latest tag) @mgates3

### 0.3.8

- [x] `spack install --test root blaspp ^openblas@0.3.8`
works

### 0.3.9

- [x] `spack install --test root blaspp ^openblas@0.3.9`
works

### 0.3.13 (latest)

Probably the most important test, because this is the *default* user experience with Spack:

- [x] `spack install --test root blaspp`
works

Just checking, because yolo:

- [x] `spack install --test root blaspp ^openblas threads=openmp`
works

### Any without locking

- [x] `spack install --test root blaspp ^openblas ~locking`
conflict added for all OpenBLAS releases >=0.3.7 (see 0.3.6 note above)

alalazo commented:
Lot of things happening overnight... 🙂 FYI, to bootstrap `clingo` there's also #21446 which "just" requires to have a compiler supporting C++14. It currently installs `clingo` in `$HOME/.spack/bootstrap` for the interpreter being used to run Spack and imports the module on the fly.

@G-Ragghianti `clingo` can be used beyond `spack solve`. You just have to set in your configuration:
```yaml
config:
  concretizer: clingo
```
and ensure Spack can import the `clingo` module. There are still a few limitations to be lifted, and it's still slower than the original concretizer (see #21289 for possible improvements) but it can already work around conflicts and take dependency into account (as you noted above) in `depends_on` directives.

ax3l commented:
> Lot of things happening overnight... :slightly_smiling_face:

Friday evenings aren't as they used to be :sweat_smile: 
Fixed a little pep8 issue now.

alalazo commented on var/spack/repos/builtin/packages/openblas/package.py:
```diff
@@ -126,6 +127,7 @@ class Openblas(MakefilePackage):
     conflicts('+consistent_fpcsr', when='threads=none',
               msg='FPCSR consistency only applies to multithreading')
 
+    conflicts('threads=pthreads', when='~locking', msg='Pthread support requires +locking')
```
Shouldn't we add also:
```python
conflicts('threads=openmp', when='~locking', msg='OpenMP support requires +locking')
```

ax3l replied:
I thought so for a moment, but the OpenBLAS docs on `USE_LOCKING` read:
```
This is done automatically for USE_THREAD=1, and should not be necessary when USE_OPENMP=1.
```

alalazo replied:
Sure, not talking about the recipe, just the spec. I think `openblas threads=openmp ~locking` won't make sense (but we can work on that in a later PR).

ax3l replied:
I think if I follow the docs that this is advertised as working, but I would not default to it reading the wording :D

alalazo reviewed (changes requested):
Basically LGTM, just a minor question.

mgates3 commented:
> No, please don't! This is an issue for downstream parallel libraries like Trilinos that assume they manage the parallelism. Our particle transport codes use MPI and OpenMPI parallelism and make BLAS calls from each process/thread; with parallel BLAS that means extra threads stepping on each other.

Oddly, this is exactly what we also want in BLAS++: that BLAS++ does the OpenMP parallelism for a batched-BLAS call, and BLAS itself is single threaded. But it seems (at least in some environments) that non-threaded OpenBLAS is not thread safe, so we get failures. Hence the conflict with openblas threads=none. Adding locking without threading would be sufficient for BLAS++. Same applies to SLATE, which is the parent project that BLAS++ was built within, that SLATE handles the parallelism and BLAS should generally be single threaded.

mgates3 commented:
>     * [x]  `spack install --test root blaspp ^openblas@0.3.7 threads=openmp`
>       same CMake error
> 
>     * [ ]  Do we want to simply add a conflict for selected old OpenBLAS releases in BLAS++'s `package.py`? Maybe that's also solved with newer BLAS++ releases that reworked the CMake, but the nwest release is still 2020.10.02 (latest tag) @mgates3

We can investigate these issues. I thought with `threads=openmp` it would be working. There have been some fixes in BLAS++ CMake, but not major CMake work since 2020.10.02, to my recollection. Most newer work has been adding ROCm support.

Mark

sethrj commented:
Interesting, I was assuming that openmp and pthreads both meant "use this threading implementation to maximize CPU utilization" rather than "be aware of downstream library's utilization of these threading implementations". For the OpenMP case, it sounds like it's the latter?

alalazo commented:
Thanks @ax3l !

ax3l commented:
Thank you all, too! Epic that we got this solved! :)

G-Ragghianti commented:
> I would even go more concrete to use
> 
> ```
> depends_on('openblas threads=openmp', when='+openmp ^openblas')
> ```
> 
> (Works well with clingo, does fail with the original dependency solver)
> 
> > I haven't tested USE_LOCKING for openblas, but I will soon.
> 
> Wuhu, thanks to the great test/check implementation in the `blaspp` package, I gave it a try already: `spack install --test root blaspp`. This is the issue I saw:
> 
> ```
> 217 tests FAILED for batch-gemm
> ...
> BLAS : Bad memory unallocation!
> ```
> 
> Indeed, enabling `USE_LOCKING=1` introduced with OpenBLAS 0.3.7+ solves the test issues! 🎉
> 
> I updated this PR to reflect the changes in OpenBLAS and BLAS++ and will wait for your and @mgates3's confirmation that this indeed is the solution.
> 
> I already did `spack install --test root blaspp` tests with OpenBLAS (Ubuntu 20.04/x86_64, GCC 10.2, original concretizer):
> ### 0.3.6
> 
>     * [x]  `spack install --test root blaspp ^openblas@0.3.6`
>       Spack abort by default with conflict (original concretizer).
>       `USE_LOCKING` came in [OpenBLAS 0.3.7 according to their docs](https://github.com/xianyi/OpenBLAS/blob/v0.3.7/Makefile.rule#L61-L65), so I added a new specific conflict
> 
>     * [x]  `spack install --test root blaspp ^openblas@0.3.6 threads=openmp`
>       BLAS++ 2020.10.02 (latest tag) CMake issue:
> 
> 
> ```
>      30       BLAS library not found.
>   >> 31    CMake Error at CMakeLists.txt:295 (message):
>      32      BLAS++ requires a BLAS library and none was found.  Ensure that it is
>      33      accessible in environment variables $CPATH, $LIBRARY_PATH, and
>      34      $LD_LIBRARY_PATH.
> ```
> 
> ### 0.3.7
> 
>     * [x]  `spack install --test root blaspp ^openblas@0.3.7`
>       same CMake error as with OpenBLAS 0.3.6
> 
> 
> Otherwise this _should_ work because of the `+locking` default that came in this OpenBLAS release.
> 
>     * [x]  `spack install --test root blaspp ^openblas@0.3.7 threads=openmp`
>       same CMake error
> 
>     * [ ]  Do we want to simply add a conflict for selected old OpenBLAS releases in BLAS++'s `package.py`? Maybe that's also solved with newer BLAS++ releases that reworked the CMake, but the nwest release is still 2020.10.02 (latest tag) @mgates3
> 
> 
> ### 0.3.8
> 
>     * [x]  `spack install --test root blaspp ^openblas@0.3.8`
>       works
> 
> 
> ### 0.3.9
> 
>     * [x]  `spack install --test root blaspp ^openblas@0.3.9`
>       works
> 
> 
> ### 0.3.13 (latest)
> 
> Probably the most important test, because this is the _default_ user experience with Spack:
> 
>     * [x]  `spack install --test root blaspp`
>       works
> 
> 
> Just checking, because yolo:
> 
>     * [x]  `spack install --test root blaspp ^openblas threads=openmp`
>       works
> 
> 
> ### Any without locking
> 
>     * [x]  `spack install --test root blaspp ^openblas ~locking`
>       conflict added for all OpenBLAS releases >=0.3.7 (see 0.3.6 note above)

HI Axel.  I've found that these cmake failures (unable to detect blas) only occur when using gcc >=10.  We will be looking into it. Thanks for finding it.

ax3l commented:
Thanks @G-Ragghianti! :)

jwillenbring commented:
I finally got back to this today. I worked through an openblas error I was having and things are at least configuring now.
